wgpu-native/src/registry.rs

```diff
@@ -1,47 +1,100 @@
+use std::marker::PhantomData;
+use std::os::raw::c_void;
 use std::sync::{Arc, Mutex};
 use std::{borrow, cmp, fmt, ops, ptr};
 
 use hal::backend::FastHashMap;
 use {AdapterHandle, DeviceHandle, InstanceHandle, ShaderModuleHandle};
 
+#[cfg(not(feature = "remote"))]
+pub(crate) type Id = *mut c_void;
+#[cfg(feature = "remote")]
 pub(crate) type Id = u32;
 
-pub(crate) struct Registry<T> {
+pub(crate) trait Registry<T> {
```

Comment:
I just used a trait here to ensure we keep both registry implementations in sync, but it's not really necessary.